### DISCUSSION TITLE
Every function app requires a storage account to operate. Except?

### DISCUSSION THREAD

**Discussion by vigouredelaruse** [Azure/azure-functions-durable-extension#2387] [Q&A] [answered]
i've risen to my natural level of incompetence with regards to how 'azure functions' use storage accounts and i have questions

the [general documentation ](https://learn.microsoft.com/en-us/azure/azure-functions/storage-considerations?tabs=azure-cli)is clear

> When creating a function app, you must create or link to a general-purpose Azure Storage account that supports Blob, Queue, and Table storage. This requirement exists because Functions relies on Azure Storage for operations such as managing triggers and logging function executions.

but the durable task extension documentation for the [mssql provider](https://github.com/microsoft/durabletask-mssql) is equally clear

> The Microsoft SQL provider for the [Durable Task Framework](https://github.com/Azure/durabletask) (DTFx) and [Azure Durable Functions](https://docs.microsoft.com/azure/azure-functions/durable/durable-functions-overview) is a storage provider that persists all task hub state in a Microsoft SQL database, which can be hosted in the cloud or in your own infrastructure.

suppose you have a function app deployment that 
* runs in arbitrary kubernetes via the keda autoscaler
* uses only say, the rabbitmq trigger to start orchestrations
* uses the mssql provider for taskhub state

apologies since this drifts into 'how the keda autoscaler works' territory. however, azure storage queue is just one of any of a set of scalers one can use with keda, implying scaling via azure storage account queues is just an [option](https://keda.sh/docs/2.6/scalers/azure-storage-queue/) 

is an azure storage account required in this highly clipped scenario?

**Comment by cgillum** [comment, marked as answer]
Azure Storage is not required in your scenario. The Azure Storage requirement applies primarily to Azure Storage-based triggers and to Function apps hosted in the Azure Functions managed service.

The product team intentionally set things up so that Function apps on Kubernetes don’t strictly require Azure Storage because many users want to run their workloads on clusters that don’t have any public internet connectivity. The general documentation may need to be updated to clarify this further.

**Reply by vigouredelaruse** [reply]
yes i see. well if the product team wants to clarify that they're leaving money on the table in this instance i will be happy to acquiesce on the matter. thanks @cgillum

**Comment by vigouredelaruse** [comment]
g'day @cgillum - i'm trying out the mssql provider sans storage account settings with [this code](https://dev.azure.com/electricrucible/the%20horseless%20newspaper/_git/the-horseless-core?path=/src/workflow/horselessfx/horselessnewspaper.workflow.fx.rabbitmq) and will one at a time some associated issues; given your earlier comments i expect misconfiguration on my part and appreciate any assistance

given triggers all similar to the below
```
        [FunctionName("TenantViewFeatureDeploymentListener")]
        public async Task ComandChannelListener(
            [RabbitMQTrigger("CommandChannel", ConnectionStringSetting = "AMQPEndpoint")] TheHorselessNewspaper.Schemas.ContentModel.ContentEntities.ContentCollection contentCollection,
            [DurableClient(ExternalClient = true)] IDurableOrchestrationClient starter,
            ILogger log)
```

with no storage account set i see this startup failure while running in visual studio sans  docker
```
[2023-02-15T15:28:33.501Z] A host error has occurred during startup operation '07d97317-04a0-4893-951c-c4ca9afcc327'.
[2023-02-15T15:28:33.503Z] Microsoft.Azure.WebJobs.Script.WebHost: Secret initialization from Blob storage failed due to missing both an Azure Storage connection string and a SAS connection uri. For Blob Storage, please provide at least one of these. If you intend to use files for secrets, add an App Setting key 'AzureWebJobsSecretStorageType' with value 'Files'.
```

vs the below nominal start with storage account set
```
[2023-02-15T15:47:43.930Z] Loading extensions from E:\src\horseless-core\src\workflow\horselessfx\horselessnewspaper.workflow.fx.rabbitmq\bin\Debug\net6.0\bin. BundleConfigured: False, PrecompiledFunctionApp: False, LegacyBundle: False
[2023-02-15T15:47:43.936Z] Script Startup resetting load context with base path: 'E:\src\horseless-core\src\workflow\horselessfx\horselessnewspaper.workflow.fx.rabbitmq\bin\Debug\net6.0\bin'.
[2023-02-15T15:47:43.962Z] Loading startup extension 'Startup'
[2023-02-15T15:47:44.405Z] Loaded extension 'Startup' (1.0.0.0)
[2023-02-15T15:47:44.474Z] Loading startup extension 'RabbitMQ'
[2023-02-15T15:47:44.507Z] Loaded extension 'RabbitMQ' (1.1.0.0)
[2023-02-15T15:47:44.516Z] Loading startup extension 'RabbitMQ'
[2023-02-15T15:47:44.544Z] Loaded extension 'RabbitMQ' (2.0.3.0)
[2023-02-15T15:47:44.549Z] Loading startup extension 'SqlDurabilityProvider'
[2023-02-15T15:47:44.578Z] Loaded extension 'SqlDurabilityProvider' (1.1.0.0)
[2023-02-15T15:47:44.583Z] Loading startup extension 'DurableTask'
[2023-02-15T15:47:44.617Z] Loaded extension 'DurableTask' (2.0.0.0)
[2023-02-15T15:47:44.624Z] Loading startup extension 'AzureStorage'
[2023-02-15T15:47:44.659Z] Loaded extension 'AzureStorage' (4.0.5.0)
[2023-02-15T15:47:44.666Z] Loading startup extension 'AzureStorageBlobs'
[2023-02-15T15:47:44.705Z] Loaded extension 'AzureStorageBlobs' (5.0.1.0)
[2023-02-15T15:47:44.712Z] Loading startup extension 'AzureStorageQueues'
[2023-02-15T15:47:44.749Z] Loaded extension 'AzureStorageQueues' (5.0.1.0)
```
given host.json
```
{
  "version": "2.0",
  "extensions": {
    "durableTask": {
      "hubName": "AHubName",
      "storageProvider": {
        "type": "mssql",
        "connectionStringName": "FunctionStorage_SQLSERVER"
      }
    },
    "rabbitMQ": {
      "port": "5672",
      "queueName": "QuorumChannel"
    }
  },

}
```
startup.cs as per
```
[assembly: FunctionsStartup(typeof(horselessnewspaper.workflow.fx.rabbitmq.Startup))]
namespace horselessnewspaper.workflow.fx.rabbitmq
{
    public class Startup : FunctionsStartup {}
}
```

active package references
```
		<PackageReference Include="Microsoft.AspNetCore.DataProtection.StackExchangeRedis" Version="6.*" />
		<PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="6.*" />
		<PackageReference Include="Microsoft.AspNetCore.Mvc.NewtonsoftJson" Version="6.*" />
		<PackageReference Include="Microsoft.Extensions.Configuration" Version="6.*" />
		<PackageReference Include="Microsoft.Extensions.Configuration.UserSecrets" Version="6.*" />
		<PackageReference Include="Microsoft.Extensions.Configuration.Abstractions" Version="6.*" />
		<PackageReference Include="Microsoft.DurableTask.SqlServer.AzureFunctions" Version="1.1.1" />
		<PackageReference Include="Microsoft.NET.Sdk.Functions" Version="4.1.3" />
		<PackageReference Include="Microsoft.Azure.Functions.Extensions" Version="1.1.0" />
		<FrameworkReference Include="Microsoft.AspNetCore.App" />
```

**Reply by davidmrdavid** [reply]
@vigouredelaruse: does this issue come up even without Durable Functions? Just trying to determine if the exception you're seeing is a result of Durable Functions, or Azure Functions. From what I gather on a quick inspection, I suspect the latter is to blame.

**Reply by davidmrdavid** [reply]
Would it be possible for you to do a quick test and validate that this error only occurs when Durable Functions is configured?

**Comment by vigouredelaruse** [comment]
hi @davidmrdavid  there are a bunch of moving parts to this equation and i think i'll cook on stabilizing some of the others before i claim something's wrong with durable functions. also it's a low priority to me to operate without a storage account and a high priority to get the mssql provider working in arbitrary kubernetes; i'll get back to you

thanks

**Comment by vigouredelaruse** [comment]
m'kay i have an orchestration generating the below good news sort of nominal log in keda/microk8s situation 
* running in various clouds 
* without using a storage account
* [rabbitmqtrigger] and orchestration triggers only
* mssql storage provider running as a user with create db privs with no prior prep on the db instance

```
8793b1c08517: Function 'LeaderElectionOrchestration (Orchestrator)' scheduled. Reason: RaiseEvent:NominateQuorumParticipant. IsReplay: False. State: Scheduled. HubName: QuoromListenerHub. AppName: . SlotName: . ExtensionVersion: 2.9.1. SequenceNumber: 178.
info: Function.QuorumAnnouncerListener[2]
      Executed 'QuorumAnnouncerListener' (Succeeded, Id=84ac5ef8-148c-423a-aa76-4e4444f3d54a, Duration=4994ms)
```

these are the advertised results and i thank you for your think time
